# INTRODUCTION TO PYTHON

### CONTENT
1. Introduction to Jupyter Notebooks 
2. Variables and data structures 
3. Conditionals
4. Lists and loops
5. Diccionaries and frequency tables
6. Functions 
7. Advanced functions
8. Object-oriented language
9. Introduction to NumPy 
10. Introduction to Pandas
11. Visualization and graph creation

## 1. Introduction to Jupyter Notebooks

<img src="Figures/Jupyter.png" alt="Drawing" style="width: 100px;"/>

Jupyter notebooks is an environment (IDE) that allows to interleave text content (to be called 'Markdown' cell) with Python code ('Code' cell) that can be executed cell by cell as shown in the following examples.

The basic commands to start using Jupyter Notebooks are:

<img src="Figures/menu-vEng.png" alt="Drawing" style="width: 850px;"/>

* `Shift+Enter`: Execute cell and select next cell.
* `Alt+Enter`: Execute cell and insert a new cell.
* `ESC`: to start 'command' mode. Once in this mode:
    * `H` list of commands
    * `A` insert cell below
    * `B` inster cell above
    * `D,D` D twice to delete cell
    * `Y` change cell type to 'Code'
    * `M` change cell type to 'Markdown'
    * `Enter` go back to 'edit' mode.

#### Execute the cell below using any of the possible commands

In [1]:
3 + 4 + 9

16

In [2]:
# Python code example
# Change values a and b, and observe the result when executing the cell 

a=87
b=34
c=a+b

print('The result of adding {} and {} is {}.'.format(a,b,c))

The result of adding 87 and 34 is 121.


## 2. Variables and data structures

Variables in Python are created once they are defined, that is, when a value is assigned to the variable for the first time (using the <code> = </code> operator). For this reason, variables (and their type) do not need to be declared in Python. 

The main types of variables that can be found are:
* **Numeric**
    * **Integer:** Positive and negative natural numbers.
    * **Float:** Real numbers.
    * **Complex:** Complex numbers. The imaginary part is multiplied by `j` to represent the root of `-1` Números complejos.
* **Boolean**
    These are variables that can take True or False as a value, represented in Python by `True` and `False` using uppercase for the initial letter.
* **Data structures (data sequences)**.
    Ordered list of values of the same or different type. In Python they exist:
    * **String:** text strings
    * **List:** ordered sets of elements
    * **Tuple:** ordered and immutable sets of elements
    * **Dictionary:** sets of elements characterized by an identifier and a value  

There are some general rules for variable names:
+ Only letters, numbers and "underscores" (_) are used.
+ NO spaces are used anywhere in the name
+ The name cannot start with a number
+ Uppercase letters are treated as different from lowercase letters (i.e., Python is case-sensitive)

In [3]:
# The name can't start with a number
3a = 10

SyntaxError: invalid syntax (173782835.py, line 2)

Whenever a variable is created or modified, Python interprets what type of variable it is. When a new value is assigned to an existing object (*dynamic typing*), the previous values of that object are cleared from the computer's memory. The first value of **a** is the real number 13.0, but after a new assignment:

In [4]:
# Dynamic typing
a = 13.0
a = 'b'
print(a)

b


The ``type()`` function displays the TYPE of the data we introduce:

In [25]:
x = True
print(type(x))

name = 'Vinicius'
print(type(name))

age = 32
print(type(age))

money = 5.5
print(type(money))

list0 = ['Vinicius', 32, 5.5]
print(type(list0))

<class 'bool'>
<class 'str'>
<class 'int'>
<class 'float'>
<class 'list'>


We can print the elements in the list or any variables using the command `print` and `.format` or simply summing strings:

In [27]:
print('My name is {}, I have {} years old and {} euros.'.format(list1[0], list1[1], list1[2]))

#or

print('My name is {}, I have {} years old and {} euros.'.format(name, age, money))

#or

print('My name is '+ name + ', I have ' + str(age) + ' years old and ' + str(money) + ' euros')

My name is Vinicius, I have 30 years old and 5.5 euros.
My name is Vinicius, I have 32 years old and 5.5 euros.
My name is Vinicius, I have 32 years old and 5.5 euros


### 2.1 Types and basic operators

The basic operators in Python are:

+ Addition (also for strings, tuples or lists): <code>a + b</code>
+ Subtraction: <code>a - b</code>
+ Multiplication (also for strings, tuples and lists): <code>a * b</code>
+ Division: <code>a / b</code>
+ Integer division: <code>a // b</code>
+ Remainder: <code>a % b</code>
+ Exponential: <code>a ** b</code>
+ Assignment: <code>=</code>, <code>-=</code>, <code>+=</code>,<code>/=</code>,<code>*=</code>, <code>%=</code>, <code>//=</code>, <code>**=</code>
+ Boolean comparisons: <code>==</code>, <code>!=</code>, <code><</code>,<code>></code>,<code><=</code>, <code>>=</code>
+ Boolean operations: <code>and</code>, <code>or</code>, <code>not</code>
+ Operations to check belongingness: <code>in</code>, <code>not in</code>
+ Operations to identify objects: <code>is</code>, <code>is not</code>
<!-- + Bitwise operators (or, xor, and, complement): <code>|</code>, <code>^</code>, <code>&</code>, <code>~</code> -->
<!-- + Left and right bit shift: <code><<</code>, <code>>></code> -->

In [17]:
# Boolean or logic operators

a = 7
b = 2

print(a>b and b<a)

print(a==7 and b==5)

print(a==7 or b==5)

True
False
True


In [18]:
# Operators to check belongingness

my_list = [1, 3, 2, 7, 9, 8, 6]
print(4 in my_list)
print(8 in my_list)

False
True


In [21]:
# Operators to identify objects

a = 7
b = 2  
c = 7
print(a is c)
print (a is not b)
print (a is not c)

True
True
False


### Python as a calculator

Python has a concise notation for arithmetic that closely resembles the traditional way of writing operations.

In [29]:
a = 3+2
b= 3.5 * -8
c = 10/6
print(a, b, c, 10./6.)

5 -28.0 1.6666666666666667 1.6666666666666667


In [30]:
# Ejemplo

a = 12
b = 13.3
c = a + b
print (c, type(a), type(b), type(c))

25.3 <class 'int'> <class 'float'> <class 'float'>


For more complex operations, we need to use other packages such as `math` and `numpy`.

In [33]:
import numpy as np   # only executed if the library hasn't been imported previously
print(np.pi)
print(np.sin(1.8))
print(np.arccos(0.8))

3.141592653589793
0.9738476308781951
0.6435011087932843


### 2.2 Data structures

### Strings or text strings

Text in Python is treated as a list of characters. Thus, the methodology that applies to the treatment of lists can be applied to text strings.

In [51]:
print("This is 'an example' of how to use quotation marks inside a text that is written between double quotation marks")
print('Here is "another example" but this time inverted')

This is 'an example' of how to use quotation marks inside a text that is written between double quotation marks
Here is "another example" but this time inverted


The '+' operator can be used to join character strings:

In [ ]:
a = 'He'
b = 'llo'
c = a+b+'!'
print(c)

To access only a part of a list or, in this case, of a *string*, the technique called **slicing** is used. The format to follow to access only a part of the string is as follows:

### <center>variable_name[first_element_index:final_index:step]</center>

Where the operation of the indexes can be seen in the following picture:

![Slicing1](Figures/slicing1.png)


In [84]:
#Example
a = 'abcdef'

print('Everything: {}'.format(a[:]))
print('First value: {}'.format(a[0]))
print('Second to end: {}'.format(a[2:]))
print('First 3: {}'.format(a[:3]))
print('Second to Fourth: {}'.format(a[2:4]))
print('All odd positions: {}'.format(a[::2]))
print('All even positions: {}'.format(a[1::2]))

Everything: abcdef
First value: a
Second to end: cdef
First 3: abc
Second to Fourth: cd
All odd positions: ace
All even positions: bdf


In the case of *strings*, there are default methods in Python that can be useful. For example the following functions:
- len: returns the length of the text.
- str: converts a number into a string.
- reverse: returns the text backwards.

We can also use string-specific methods:
- lower: converts all text to lowercase.
- upper: converts all text to uppercase.
- capitalize: capitalizes the first letter of the string.
- split(delimiter): returns a list of substrings separated by the specified delimiter (see examples below).


In [92]:
text='We are going to test the Strings Functions and Methods!'
print(len(text))
print(text.lower())
print(text.split(' '))

######################################################################3
# Return the 4th word of the text in all capital letters
print(text[7:12].upper())

55
we are going to test the strings functions and methods!
['We', 'are', 'going', 'to', 'test', 'the', 'Strings', 'Functions', 'and', 'Methods!']
GOING


### Lists

Lists are a Python type used to group objects of other types (even lists!). For example a list of numbers, of floats, of strings, of other lists, and so on. 

Lists, like strings, can use the *slicing* methodology.


In [94]:
# Create list
my_list = [0,5,'6',34,'Hello']
my_list

[0, 5, '6', 34, 'Hello']

In [95]:
# Access to an element of a list
a = my_list[2]
a

'6'

In [96]:
# Change an element of a list
my_list[-1]='Bye'
my_list

[0, 5, '6', 34, 'Bye']

In [97]:
# We can have lists of lists
my_list=[45,[3,5,6],'word',34.6,['a','b','c']]
print(my_list)
print(my_list[1])
print(my_list[1][2])
my_list[4][2]

[45, [3, 5, 6], 'word', 34.6, ['a', 'b', 'c']]
[3, 5, 6]
6


'c'

For lists there are also default Python functions and methods that can be useful. For example the following functions:
- len: returns the length of the list.
- list: create list

To use a function, call function_name(variable_name).

We can also use list-specific methods:
- append: add element to the end of the list.
- insert: add an element in a particular position.
- pop: remove the last element of the list.
- remove: remove a particular element.
- sort: sort the list

To use a method, call variable_name.method_name.

In [99]:
my_list=['first','hello','test','bye','animals','master']
print(my_list)

my_list.append('last')
print(my_list)

my_list.insert(1,'second')
print(my_list)

my_list.pop()
print(my_list)

my_list.remove('test')
print(my_list)

my_list.sort()
print(my_list)

['first', 'hello', 'test', 'bye', 'animals', 'master']
['first', 'hello', 'test', 'bye', 'animals', 'master', 'last']
['first', 'second', 'hello', 'test', 'bye', 'animals', 'master', 'last']
['first', 'second', 'hello', 'test', 'bye', 'animals', 'master']
['first', 'second', 'hello', 'bye', 'animals', 'master']
['animals', 'bye', 'first', 'hello', 'master', 'second']


### Dictionaries

A dictionary is another way of grouping data where an element can be accessed through its *key*. For example:

In [ ]:
this_installation = {
    'Power': 10,
    'Type': 'Solar',
    'Construction_year':2012
}

In [ ]:
this_installation['Type']

Another example could be:

In [102]:
dict = {"Installation_1": "Solar", "Installation_2": "Wind", "Installation_3": "Wind", "Installation_4": "Hidraulyc",
        "Installation_5": "Solar",}
print(dict)

{'Installation_1': 'Solar', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}


To access, modify or add an item, it can be done in a similar way to the lists:

In [103]:
# Access
dict["Installation_1"]

'Solar'

In [104]:
# Modify
print(dict)
dict["Installation_1"]="Wind"
print(dict)

{'Installation_1': 'Solar', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}
{'Installation_1': 'Wind', 'Installation_2': 'Wind', 'Installation_3': 'Wind', 'Installation_4': 'Hidraulyc', 'Installation_5': 'Solar'}


In [ ]:
# Add
print(dict)
dict["Installation_6"]="Solar"
print(dict)

Dictionaries can contain all types (int, floats, lists, other dictionaries, etc.).

In [111]:
dict = {"Installation_1": ["Solar", '10 MW'], "Installation_2": ["Wind", '15 MW'], "Installation_3": ["Solar", '18 MW'],
        "Installation_4": ["Hydraulic", '12 MW'], "Installation_5": ["Solar", '23 MW'], "Installation_6": ["Wind", '8 MW'],
        "Installation_7": ["Solar", '11 MW'], "Installation_8": ["Hydraulic", '11 MW']}
dict

{'Installation_1': ['Solar', '10 MW'],
 'Installation_2': ['Wind', '15 MW'],
 'Installation_3': ['Solar', '18 MW'],
 'Installation_4': ['Hydraulic', '12 MW'],
 'Installation_5': ['Solar', '23 MW'],
 'Installation_6': ['Wind', '8 MW'],
 'Installation_7': ['Solar', '11 MW'],
 'Installation_8': ['Hydraulic', '11 MW']}

In [113]:
dict['Installation_3'][1]

'18 MW'

In [ ]:
##########################################################
# Change power of installation 4 to 15 MW modifying the dictionary!
dict['Installation_4'][1]='15 MW'
print(dict['Installation_4'])

### Tuples

Tuples are lists **that cannot be modified.**

Their syntax is the same as lists, but to write a tuple you use parentheses and not square brackets.

In [100]:
mysupertuple = ('Solar', 'Wind', 'Hydraulic', 'Gas', 'Nuclear', 'NoMoreCoal')
print(type(mysupertuple), mysupertuple[4:])

<class 'tuple'> ('Nuclear', 'NoMoreCoal')


In [101]:
mysupertuple[0]='PV'
#CANNOT BE MODIFIED!

TypeError: 'tuple' object does not support item assignment

## 3. Conditional structures (if)

Conditional control structures allow you to execute one part of the code or another depending on the evaluation of one or several boolean conditions of YES or NO (**<code>TRUE</code>** o **<code>FALSE</code>**).

In Python, conditional control structures are defined by the words **<code>if</code>**, **<code>elif</code>** y **<code>else</code>**.

- **<code>if CONDICION:</code>** if the conditional expression is met, the following code block is executed.
- **<code>elif CONDICION:</code>** otherwise, if this conditional expression is met, this other code block is executed.
- **<code>else:</code>** otherwise, this code block is executed without evaluating any condition.


The condition is usually evaluated by the following relational operators <code> <, <=, ==, >=, >, != </code>.

The part of code inside an **<code>if</code>** cannot be empty, but if for some reason an if statement must be created without content, the **<code>pass</code>** statement can be used to prevent an error from occurring.

In [119]:
# Example

celsius = 0
fahrenheit = 9.0 /5.0 * celsius + 32

print("Temperature in Fahrenheit is", fahrenheit)

if fahrenheit > 90:
    print("It is hot")
elif fahrenheit < 30:
    print("It is cold")
else: print("It is neither hot nor cold")

Temperature in Fahrenheit is 32.0
It is neither hot nor cold


#### Exercise

Create a code to identify whether the variable **<code>number</code>** is a positive, negative, or 0 value. Store the result in a variable called **<code>sign</code>**.

In [ ]:
number = 2

###############################
if number > 0:
    sign = 'positive'
elif number < 0:
    sign = 'negative'
else:
    sign = 'zero'

print(sign)

The conditional structure can also be expressed in a single line as follows, depending on whether one, two or more conditions are involved:

In [ ]:
#Example

a = 330
b = 330

# with one only condition:
if a > b: print("a is larger than b")
    
# with two conditions:
print("a is larger") if a > b else print("a is not larger")

# with 3 conditions:
print("A") if a > b else print("=") if a == b else print("B")

## 4. Lists and loops

### 4.1 Iterative control structures (for & while)

Iterative control structures (also called loops) allow the same code to be executed repeatedly as long as a condition is met.

There are two types of iterative control structures:

* For
* While

The Python ``for`` statement iterates over the elements of any sequence (a list or a string for example), in the order they appear in the sequence.

The ``while`` loop allows multiple iterations to be performed based on the result of a logical expression that can result in a ``True`` or ``False`` value.

In [121]:
# Example of for

for i in range(0,10,1):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [122]:
# Example of while

i=0
while i < 10 :
    print(i)
    i= i+1

0
1
2
3
4
5
6
7
8
9


In addition, loops with ``for`` or ``while`` can be combined with conditional ``if`` structures to create more complex structures.

With the ``break`` statement we can stop the loop before it has gone through all elements.

With the ``continue`` statement we can stop the current iteration of the loop (in both ``for`` and ``while``) and continue with the next one.

In [124]:
# Exercise: calculating the average value of a list with a for

numbers=[1,2,3,4]

#####################

sum_numbers = 0
for n in numbers:
    sum_numbers += n

average = sum_numbers/len(numbers)
print(sum_numbers, average)

10 2.5


### 4.2 Lists (o dictionary) comprehensions

List comprehensions are a way of fitting a ``for`` loop, an ``if`` statement, and an assignment all on a single line.

A list comprehension consists of the following parts:

+ An input sequence
+ A variable representing members of the input sequence
+ An optional expression
+ An output expression that produces elements of the output list from members of the input sequence that satisfy the predicate

In [125]:
# Example
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = []
for i in num:
     if i > 0:
        squared.append(i**2)
print(type(squared), squared)

# This for can be rewritten as follows:
# In an iteration, it first checks if the if is TRUE and then performs the calculation
squared = [x**2 for x in num if x > 0] 
print(type(squared), squared)

<class 'list'> [1, 16, 100, 4, 9]
<class 'list'> [1, 16, 100, 4, 9]


In [126]:
# Exercise: calculating the multiplicative inverse (1/x) of all the elements of a list with a single line for

######################################################
num =[1,2,3,4]

inverse = [1/x for x in num if x!=0]
inverse

[1.0, 0.5, 0.3333333333333333, 0.25]

This compact way of expressing loops can also be used to iterate within lists. Below are different ways to join the values of two lists using zip or nested fors.

In [127]:
list1 = ['a','b','c','d']
list2 = [1,2,3,4]

# Create a list with nested fors:
print ([(i,j) for i in list1 for j in list2])

# Create a list using zip:
print ([(i,j) for i,j in zip(list1,list2)])

# Cereate a dictionary using zip:
print ({i:j for i,j in zip(list1,list2)})

[('a', 1), ('a', 2), ('a', 3), ('a', 4), ('b', 1), ('b', 2), ('b', 3), ('b', 4), ('c', 1), ('c', 2), ('c', 3), ('c', 4), ('d', 1), ('d', 2), ('d', 3), ('d', 4)]
[('a', 1), ('b', 2), ('c', 3), ('d', 4)]
{'a': 1, 'b': 2, 'c': 3, 'd': 4}


## 6. Functions
A function is a block of code with an associated name, which receives zero or more arguments as input, follows a sequence of statements where the desired operations are executed and returns a value and/or performs a task. This code block can be called multiple times as needed.

In Python there are a number of functions built into the language by default, but you can also create user-defined functions to use in your own programs. 

The complete list of default built-in functions can be found at:
www.w3schools.com/python/python_ref_functions.asp

In [ ]:
# Example of an integrated function in Python

# The input() function allows us to assign a value entered by the user to a variable

a = input()

print('The variable introduced has been ' + a)

To create a function in Python, you must use the word ``def`` to define the function. The syntax for a function definition in Python is:

<code>*def* function_name(input_parameter_list):
    function_statements
    *return* [expression_to_return]</code>

Where:
* function_name: is the name of the function.
* input_parameter_list: is the list of parameters that a function can receive, separated by comma.
* function_statements: is the block of statements in Python source code that perform a given operation.
* expression_to_return: is the expression or variable that returns the return statement.

In [ ]:
# Example: function to set the first value of a list to 0

def change_first_element(list):
    list[0]=0 #it returns none!

numbers=[1,2,3,4]
change_first_element(numbers)
print(numbers)

#### Example: greatest common denominator (GCD)

The common denominator of two positive integers $a$ and $b$ is the greatest common divisor between $a$ and $b$ . The Euclidean algorithm is an iterative method for computing the greatest common denominator of two integers. In pseudocode it would be:

+ If $a<b$, change $a$ and $b$.
+ Divide $a$ by $b$ and obtain the residue, $r$. 
+ If $r=0$, the value $b$ is GCD of $a$ and $b$.
+ If $r \neq 0$, iterate again replacing $a$ by $b$ and replacing $b$ by $r$. 

In [ ]:
def GCD(a,b):
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

GCD(100,16)

#### Exercise: create a function to calculate the factorial of an integer.

The factorial of a non-negative integer $n$ is the product of all positive integers less than or equal to $n$.

In [ ]:
#########################################

def fact_1(n):
    factorial_total = 1
    while n > 1:
        factorial_total *= n
        n -= 1
    return factorial_total

fact_1(5)

# from math import factorial
# factorial(5)

### 6.1 Arguments and parameters of functions

When defining a function, the values that are received are called parameters, while during the call, the values that are sent are called arguments.

By default, a function must be called with the correct number of arguments. If the function expects 2 arguments, you must call the function with 2 arguments.

#### Arguments per position

When arguments are sent to a function, by default they are received in order in the defined parameters, known as arguments by position. In the example above (GCD), argument 100 is position 0, which corresponds to the parameter of function a, while argument 16 is position 1, which corresponds to the parameter of function b.

#### Arguments by name

However, it is possible to skip the order of the parameters if you indicate during the call what value each parameter has from its name.

In [ ]:
GCD(b=100,a=16)

#### Default parameters

To avoid error messages when a function is not called with the required arguments, you can define default parameters within a function.

In [ ]:
def GCD(a=None,b=None):
    if a == None or b == None:
        print ("Error, you must send two numbers to the function.")
        return
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

GCD()

#### ndeterminate parameters ``*args`` y ``**kwargs``

Sometimes it is not known in advance how many items need to be sent to a function.

If you do not know how many arguments per item will be passed to a function, you add a * before the parameter name in the function definition. By convention, the parameter name ``*args`` is usually used. In this way, the function will receive a tuple of arguments and can access the items accordingly.

Similarly, if it is not known how many arguments by name will be passed to a function, two ** are added before the parameter name in the function definition. By convention, the parameter name ``**kwargs`` is usually used. In this way, the function will receive a dictionary of arguments and can access the elements accordingly.

In [120]:
def super_function(*args,**kwargs):
    total = 0
    for arg in args:
        total += arg
    print ("sum => ", total)
    for kwarg in kwargs:
        print (kwarg, "=>", kwargs[kwarg])

super_function(50, -1, 1.56, 10, 20, 300, name="Peter", age=38)

sum =>  380.56
name => Peter
age => 38


## 9. Introduction to NumPy

NumPy is a Python library used to work with matrices. It also has functions for working with linear algebra or Fourier transforms. NumPy stands for Numerical Python.

<img src="Figures/NumPy_logo.png" alt="Drawing" style="width: 400px;"/>

As we have seen, in Python we already have lists that serve the purpose of working with arrays, but their processing is slow. NumPy aims to provide an array object that is up to 50 times faster than traditional Python lists, something important for data science where speed and resources are very important. In addition, they differ from lists in that NumPy arrays cannot be resized, all elements must be of the same type, and inter-array operations are allowed.

The NumPy source code can be found in this github repository https://github.com/numpy/numpy

The array object in NumPy is called ``ndarray``, it provides many support functions that make working with ``ndarray`` very easy. You can create an ndarray with NumPy with the ``.array()`` function.

In [ ]:
#Example
import numpy as np #normally imported as np

arr = np.array([1, 2, 3, 4, 5])

print(arr)

print(type(arr))

To create an ndarray, we can use a list, tuple or any array object to the ``array()`` method, and it will become an ``ndarray``. You can create ``ndarrays`` of different dimensions. 0-D are just numbers, 1-D are vectors, 2-D are arrays, and so on.

The main attributes that can be used with ``ndarrays`` are:

| Attribute | Description | 
|-----------|-------------|
| ndim |number of dimensions of the matrix| 
| shape |number of elements of each dimension of the array|
| size |total number of elements in the matrix|
| dtype | type of the elements of the matrix|
| astype |to change the type of the elements of the matrix|

In [ ]:
#Example

a = np.array(42)
b = np.array([1, 2, 3, 4, 5])
c = np.array([[1, 2, 3], [4, 5, 6]])
d = np.array([[[1, 2, 3], [4, 5, 6]], [[0, 1, 2], [3, 4, 5]]])

print(a.ndim)
print(b.ndim)
print(c.ndim)
print(d.ndim)
print(d.shape)
print(d.size)
print(d.dtype)

e = d.astype(float)

print(e)
print(e.dtype)

### 9.1 Accessing values in an ndarray

An array element can be accessed by referring to its index number. Indexes in NumPy arrays start with 0, which means that the first element has index 0, and the second has index 1, etc.

To access elements of 2-D arrays we can use comma-separated integers representing the dimension and index of the element. For 3-D arrays or higher, you must add as many comma-separated integers as there are dimensions in the array.

In [ ]:
#Example

arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print(arr[0, 1])

Slicing can also be used in ``ndarrays``. In this case: ``array[indice_primer_elemento:indice_final:step]``

In [ ]:
# Exercise

arr = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])

# returns the index 2 of the two dimensions:

arr[:, 2]

# returns the interval between index 1 and 4 (not included) of the two dimensions:
arr[:, 1:4]

# devuelve todos los elementos de posiciones pares:
arr[:, ::2]

<img src="Figures/numpy_indexing.png" alt="Drawing" style="width: 200px;"/>

In [ ]:
# Exercise: select the elements of each of the sets of elements in the previous figure:

arr = np.array([[0,1,2,3,4,5],[10,11,12,13,14,15],[20,21,22,23,24,25],[30,31,32,33,34,35],[40,41,42,43,44,45],[50,51,52,53,54,55]])
arr

# New array with selection of red elements:
arr2 = arr[:,2]
print(arr2)
# New array with selection of orange elements:
arr3 = arr[0, 3:5]
print(arr3)
# New array with selection of green elements:
arr4 = arr[2::2, ::2]
print(arr4)
# New array with selection of blue elements:
arr5 = arr[-2:, -2:]
print(arr5)

### 9.2 Importing arrays from .csv files

Since **Comma-separated values (CSV)** files are a .TXT file type, you can import a .CSV file with the ``np.genfromtxt(file_location)`` function.

The optional parameters of **genfromtxt** can be found in: https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html

| Function | Type | Description | 
|-----------|----|---------|
|skip_header | optional | number of lines to avoid at the beginning of the file|
|filling_values | optional |values to be used when there are no values|
|delimiter | optional |string used to separate values (in CSV it is ',')|
|usecols | optional | selects which columns to import, with 0 being the first (e.g. ``usecols = (1, 4, 5)``)|


#### Exercise: 

Look at the file **Data/block_13_daily_reduced.csv** and see how it has been imported into NumPy.

Replace nan values with ``99999`` and avoid importing the first row and the first two columns.

In [ ]:
data = np.genfromtxt('Data/block_13_daily_reduced.csv',delimiter=',')
print (data)

##########################

# Replace nan values with 99999
data1 = np.genfromtxt('Data/block_13_daily_reduced.csv',delimiter=',', filling_values=99999)
print (data1)


# Avoid importing the first row and the first two columns
data2 = np.genfromtxt('Data/block_13_daily_reduced.csv', delimiter=',', skip_header= 1, usecols=(2,3,4,5,6,7,8))
print (data2)

### 9.3 Basic functions with NumPy

Here you can see a list of the most interesting functions to look at in the documentation, with some examples:

| Function | Description | 
|-----------|-------------|
| array_name.mean() |Average| 
| array_name.min() |Minimum value|
|| array_name.max() |Maximum value|
| | array_name.sum() |Sum of values|
| array_name.std() |Standard deviation|
| array_name.var() |Variance|
| array_name.reshape(new_dimension/en) |Allows to change the dimensions|

Some examples of Methods associated with NumPy:

| Method | Description | 
|-----------|-------------|
| np.amin(array_name) | minimum value |
| np.amax(array_name) | maximum value|
| | np.argmax(array_name) | maximum value index|
| np.argmin(array_name) | index of the minimum value| | np.argmin(array_name) | index of the minimum value|
| np.isnan(array_name) | value identifier Nan|

For a multidimensional matrix, it is possible to do the calculation along a single dimension by passing the ``axis`` parameter to indicate the axis. For example, in 2-D, axis=0 is the column and axis=1 is the row.

#### Exercise

Practice with some of these functions or Methods from the ``data`` array.

In [ ]:
# Your code here!

################################################

# Find value and position of the maximum value
max_value = data2. max()
print('Maximum value', max_value)

# Find value and position of the minimum value of the second column
min_value = data2. min()
print('Minimum value', min_value)

# Find average value of each row
mean = list(map(lambda x: x.mean(), data2))
print('Mean value', mean)

# Find position of Nan values
np.isnan(data, where=True)

### 9.4 Filtering data with a Boolean array

Boolean comparisons can be used to compare elements in arrays of equal size.

The ``where`` function creates a new array from comparing two arrays, following the following syntax: ``where(bool_array, true_array, false_array)``

In [ ]:
# Example: avoid error when dividing by zero

a = np.array([1, 3, 0], float) 
np.where(a != 0, 1/a, 0) 

This can be used to create a mask by extracting the indices of an array that satisfy a given condition.

In [ ]:
arr = np.array([10,8,30,40])
print (arr)
mask = arr < 9 # boolean array with elements less than 9
mask

#### Exercise

Rewrite the following code to obtain an array where values less than 9 are set to 10 using the previous mask array and the ``where`` function.

In [ ]:
print ('Resetting all values below 9 to 10...')

##############################
print(np.where(arr < 9, 10, arr))

### For more information about NumPy

* http://numpy.scipy.org
* http://scipy.org/Tentative_NumPy_Tutorial
* https://numpy.org/learn/

## 10. Introducción a Pandas

**Pandas is one of the most important libraries in Python for Data Science!!!!** Basically it will allow us to have ``Datasets`` opened as if they were an excel spreadsheet, that is, in matrix form with two dimensions and with indexed columns and indexes.

<img src="Figures/Pandas.jpg" alt="Drawing" style="width: 800px;"/>


Pandas is used to:
* Operate on the whole dataset or on a row or column (vectors) with high computational efficiency
* Helps to clean data and fix missing data
* Access subsets of data
* Add or remove columns with new features.
* Group data by characteristics.
* Very efficient in joining data from different sources.
* Work with time series.

### 10.1 Read data from a csv

Most of the time we will load the files through a .csv file. Below you can see the basic code that will be used most of the time:

In [128]:
import pandas as pd
london = pd.read_csv('Data/block_13_diario.csv')
london.head(5)

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134


### 10.2 Data inspection

If we want to know what dimensions this DataFrame has, we can look at its 'shape'.

In [129]:
london.shape

(32992, 9)

There are 32992 entries and 9 characteristics for each entry.

It can be very useful to view the first or last entries of the dataset to better understand this data:

In [130]:
london.head(10)

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
0,MAC000113,2011-12-14,1.0910,1.065250,1.674,28,0.305397,29.827,0.262
1,MAC000113,2011-12-15,0.3710,0.546521,1.434,48,0.410120,26.233,0.119
2,MAC000113,2011-12-16,0.3205,0.534708,2.034,48,0.464431,25.666,0.129
3,MAC000113,2011-12-17,0.2570,0.421500,1.068,48,0.274387,20.232,0.115
4,MAC000113,2011-12-18,0.3985,0.565333,2.218,48,0.450712,27.136,0.134
5,MAC000113,2011-12-19,0.6980,0.746479,2.204,48,0.488694,35.831,0.131
6,MAC000113,2011-12-20,0.2485,0.436354,1.551,48,0.362781,20.945,0.139
7,MAC000113,2011-12-21,0.2050,0.426542,1.818,48,0.477511,20.474,0.119
8,MAC000113,2011-12-22,0.1830,0.181854,0.259,48,0.040321,8.729,0.119
9,MAC000113,2011-12-23,0.1860,0.185646,0.268,48,0.040754,8.911,0.121


In [131]:
london.tail()

,LCLid,day,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
32987,MAC005416,2014-02-24,0.0660,0.111437,0.809,48,0.126465,5.349,0.050
32988,MAC005416,2014-02-25,0.0645,0.072521,0.238,48,0.030756,3.481,0.051
32989,MAC005416,2014-02-26,0.0675,0.107917,0.343,48,0.080074,5.180,0.049
32990,MAC005416,2014-02-27,0.0910,0.104250,0.296,48,0.054844,5.004,0.047
32991,MAC005416,2014-02-28,0.0630,0.063000,0.063,1,NaN,0.063,0.063


If we want to know the name of the columns or indexes, we can access them:

In [132]:
london.columns

Index(['LCLid', 'day', 'energy_median', 'energy_mean', 'energy_max',
       'energy_count', 'energy_std', 'energy_sum', 'energy_min'],
      dtype='object')

In [133]:
london.columns[4]

'energy_max'

In [134]:
london.index

RangeIndex(start=0, stop=32992, step=1)

Finally, to have a very good summary, it is very useful to use the ``describe`` method.

In [135]:
london.describe()

,energy_median,energy_mean,energy_max,energy_count,energy_std,energy_sum,energy_min
count,32992.000000,32992.000000,32992.000000,32992.000000,32865.000000,32992.000000,32992.000000
mean,0.165632,0.223007,0.879596,47.776128,0.182819,10.660024,0.064419
std,0.158396,0.178484,0.672407,3.035524,0.147394,8.565058,0.076632
min,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000
25%,0.052000,0.085432,0.348000,48.000000,0.069162,4.046000,0.016000
50%,0.118500,0.176844,0.775000,48.000000,0.149607,8.458000,0.035000
75%,0.225000,0.311276,1.248000,48.000000,0.264400,14.910500,0.093000
max,1.608000,1.407458,6.897000,48.000000,1.104101,67.558000,1.068000


### 10.3 Select data

If we want to work or select only a part of the dataset, we can use **[ ]** after the dataset name, and indicate which columns we want to select.

In [ ]:
london[['day', 'energy_max']].head()

If, on the other hand, you want to select a part of the rows, it will be indicated using *slicing* as in the strings:

In [ ]:
london.iloc[0]

#### loc vs iloc

To select more specific parts (more than one column, several columns and only a few rows of these,...), indexing with loc or iloc will be used. The only difference between the two is that the first one is based on the *label* of the rows or columns, while iloc is based on the *integer* of these. With an example it is better understood:

In [ ]:
london.loc[3:5,['day', 'energy_max']]

In [ ]:
london.iloc[3:5,1:3]

Note that the index will not always be a number and can be another parameter. For example, we can set the day as the index.

### 10.4 Filtering data

Another way to select only a part of the dataset is to apply *Boolean indexing*, which is nothing more than applying a filter to the data. For example, we will select only those days in which the maximum energy has been greater than 1.

In [ ]:
london['energy_max'] > 1

As you can see, now we have selected those columns that meet our condition. What if we put it directly inside the dataset when we select only one part?

In [ ]:
london[london['energy_max'] > 1]

# Here we see that all the rows are no longer displayed

# london[london['energy_max'] > 1.5]

#### Exercise

Obtain a summary table of the parameters for the counter with ID 'MAC000113'

In [ ]:
########################################################################
# Your code here!
# Obtain a DataFrame with only the values of the counter with ID 'MAC000113'

# london[london['LCLid']=='MAC000113']

In [ ]:
##########################################################################
# Use the method .describe() to obtain the summary table of the parameters

# london[london['LCLid']=='MAC000113'].describe()

# We see that there is some missing value in one of the columns. This can also be checked with the method .isna().sum()

# london[london['LCLid']=='MAC000113'].isna().sum()

### 10.5 Interesting basic functions

Here you can see a list of the most interesting functions to look at in the documentation, with some examples:

| Function | Description |
|-----------|---------------|
| count() |Number of non-null observations|  
| sum() |Sum of values|
| mean() |Average of observations | 
| median() |Median of observations |
| min() |Minimum value|
| max() |Maximum value|
| std() |Standard deviation|
| var() | Variance|
| value_counts() | Unique values in each column - freq table|
| nunique() | Number of distinct values|
|  isnull() | Mask for null or Nan|

Practice with some of these functions.

In [ ]:
##############################################################
# your code here!

# Total energy value for ID 'MAC000113'

id_MAC000113 = london[london['LCLid']=='MAC000113']
print(id_MAC000113['energy_sum'].sum())

# How many different counters are there and what are their names?

london['LCLid'].value_counts()

In [ ]:
#####################################################################
# What was the maximum (hourly) energy value on May 15, 2013?

london[london['day']=='2013-05-15']

london[london['day']=='2013-05-15']['energy_max'].max()

In [ ]:
###################################################################################
# What was the maximum daily energy consumption value on May 15, 2013?

london[london['day']=='2013-05-15']['energy_sum'].max()

### 10.6 Manipulating dataframes

You can apply a function to some column using the *apply* method and remembering the lambda functions!

In [ ]:
london['energy_var'] = london['energy_std'].apply(lambda x: x**2)
london.head()

#### Exercise

Create a new column named 'ID' with only the last 6 digits of 'LCLid'.

In [ ]:
##############################################################
# your code here!

london['ID']=london['LCLid'].apply(lambda x: x[-6:])

In [ ]:
london.head()

### 10.7 Sort

If you want to sort values, you can use ``sort_values``. With an example it's better understood:


In [ ]:
london.sort_values(by='energy_max', ascending= False, inplace=True)
london.head()

### 10.8 Group

Finally, a very useful function is the group function (group_by).

For example, imagine that you want to group all the data by counter number and take some parameter from this counter (maximum, minimum, etc.).

The result of this operation will be another dataframe with the data grouped according to a variable. In addition, you must also indicate which *aggregation* function you want to use (max, min, mean, count, sum, etc.).

Let's see an example!

In [ ]:
smart_meters = london[['LCLid','energy_max', 'energy_min', 'energy_sum']].groupby('LCLid').count()
smart_meters

### 10.9 Merge

Often we will have the desired information in different databases, files or datasets. Pandas offers us the option to join and merge datasets with *merge*.

Let's see an example with the dataset *london*. 

If we load another dataset with the information of each house we obtain:

In [ ]:
import pandas as pd

london = pd.read_csv('Data/block_13_daily.csv')
london.head(5)

In [ ]:
houses = pd.read_csv('Data/informations_households.csv')
houses.head(5)

We can see how for each counter, it gives us information about where we can find more parameters of that counter and groups them by groups called Acorn. To know more about these, go directly to the repository of this dataset and inspect in detail.

Anyway, if we want to gather all the information in the same dataset, we would have to make a merge:

<img src="Figures/Panda Data Wrangling Cheat Sheet 2.jpeg" alt="Drawing" style="width: 500px;"/>


Where the possible *how* modes refer to:

<img src="Figures/merge_options.png" alt="Drawing" style="width: 500px;"/>


In [ ]:
combined = pd.merge(london, houses, how='left', on='LCLid')
combined.head(5)

### 10.10 Other

There are many more functions that can be used with Pandas. We encourage you to take a look at their documentation, forums, cheat sheets, courses,...

https://pandas.pydata.org/pandas-docs/stable/index.html

## 11. Graphics creation

In Python there are several libraries to create graphics such as Bokeh, Altair, Pandas or Matplotlib.

<img src="Figures/four_logos.png" alt="Drawing" style="width: 800px;"/>

In this course we will learn the basic functions of matplotlib, since the other libraries are always based on it and it will help us to know how to use them as well.

### 11.1 Matplotlib.pyplot

The module used is pyplot from matplotlib. All its documentation can be found here:

https://matplotlib.org/api/pyplot_api.html#module-matplotlib.pyplot

Let's import it!

In [ ]:
import matplotlib.pyplot as plt

The two most important instances in pyplot are
* figure: in matplotlib.figure.Figure, a *figure* is where the graphics will go. It is the canvas where one or more graphics can be placed.
* axes: in matplotlib.axes.Axes, we find where the actual graphics will go.

Let's create a first test graph:

In [ ]:
import matplotlib.pyplot as plt
import numpy as np

plt.plot([1,2,3,4,5,6,7,8,9,10],[1,4,8,16,25,36,49,64,81,100])
plt.show()

In this case we have created a plot using *plot*, where the first argument is the x-axis and the second the y-axis (they must be of the same dimensions). 

We can add other information such as title and name of the axes:

In [ ]:
import matplotlib.pyplot as plt

plt.plot([1,2,3,4],[1,4,8,16])
plt.title('Our first plot!')
plt.xlabel('X axis')
plt.ylabel('Y axis')
plt.show()

### 11.2 Customize plots

If you want to change the size of the chart, this is where you will have to use the term *figure* discussed above. We remember that *figure* was the canvas where we put one or more graphics. In this case, we only put one.

Here we can see how to set up a *figure* with some predetermined measures:

In [ ]:
plt.figure(figsize=(20,5))
plt.plot([1,2,3,4],[1,4,8,16])
plt.show()

For each chart you can also determine the style of the chart with a third argument:

In [ ]:
plt.plot([1,2,3,4],[1,4,8,16],'r^')
plt.show()

# https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot.html

# character	description
# '.'	point marker
# ','	pixel marker
# 'o'	circle marker
# 'v'	triangle_down marker
# '^'	triangle_up marker
# '<'	triangle_left marker
# '>'	triangle_right marker
# '1'	tri_down marker
# '2'	tri_up marker
# '3'	tri_left marker
# '4'	tri_right marker
# '8'	octagon marker
# 's'	square marker
# 'p'	pentagon marker
# 'P'	plus (filled) marker
# '*'	star marker
# 'h'	hexagon1 marker
# 'H'	hexagon2 marker
# '+'	plus marker
# 'x'	x marker
# 'X'	x (filled) marker
# 'D'	diamond marker
# 'd'	thin_diamond marker
# '|'	vline marker
# '_'	hline marker

And we can put more than one graphic in the same plot, indicating in the legend what each thing is:

In [ ]:
plt.plot([1,2,3,4],[1,4,8,16],'g-',label='first')
plt.plot([1,2,3,4],[1,6,12,6],'r-', label='second')
plt.title('Now we have two graphics!')
plt.xlabel('X axis')
plt.ylabel('Y axis')
plt.legend()
plt.show()

### 11.3 Subplots

To represent different plots with different axes you can use Subplots. The subplot method has 3 arguments: number of rows, number of columns and index. Let's see with an example how to use it:

In [ ]:
plt.figure(figsize=(15,2))

plt.subplot(1,2,1)
plt.plot([1,2,3,4],[1,4,9,16],'go')
plt.title('Left')
plt.xlabel('X Left')
plt.ylabel('Y Left')

plt.subplot(1,2,2)
plt.plot([1,2,3,4],[1,6,12,6],'r-')
plt.title('Right')
plt.xlabel('X Right')
plt.ylabel('Y Right')

Or, for example:

In [ ]:
plt.figure(figsize=(10,10))

plt.subplot(2,1,1)
plt.plot([1,2,3,4],[1,4,9,16],'go')
plt.title('Above')
plt.xlabel('X above')
plt.ylabel('Y above')

plt.subplot(2,1,2)
plt.plot([1,2,3,4],[1,6,12,6],'r-')
plt.title('Below')
plt.xlabel('X Below')
plt.ylabel('Y Below')

### 11.4 Bar chart

To make a bar chart instead of `plot` we will use `bar`, following a very similar structure.

Recall the dataset from before:

In [ ]:
meters = london[['LCLid','energy_max', 'energy_min', 'energy_sum']].groupby('LCLid').mean()
meters.head(10)

We can make a bar chart:

In [ ]:
divisions = meters.index[:10]
marks = meters['energy_max'][:10]

plt.figure(figsize=(20,10))
plt.bar(divisions, marks, color = 'blue')
plt.xlabel('Meter')
plt.ylabel('Maximum Energy')
plt.xticks(rotation=45, ha='right')

Or in horizontal bars, using `barh`:

In [ ]:
divisions=meters.index[:10]
marks=meters['energy_max'][:10]

plt.figure(figsize=(20,10))
plt.barh(divisions, marks, color = 'blue')
plt.ylabel('Meter')
plt.xlabel('Maximum Energy')

### 11.5 Histogram

Let's see with an example how to make a histogram:

In [ ]:
plt.hist([1,5,8,7,4,1,2,4,3,6,5,4,6,3,2,3,6,9,8,5,4,1,6,9,8,7,8,8,8,7,4,7,8,8,9,9,8,7,8,9,6,9,8,7,8,9,6,6,3,2,9,8,1,9,4,1,5,6,5,6,9,6,5,8,5,6,9,6,3,2,1,4,5,8,7,8,9],9)


In [ ]:
plt.hist([1,5,8,7,4,1,2,4,3,6,5,4,6,3,2,3,6,9,8,5,4,1,6,9,8,7,8,8,8,7,4,7,8,8,9,9,8,7,8,9,6,9,8,7,8,9,6,6,3,2,9,8,1,9,4,1,5,6,5,6,9,6,5,8,5,6,9,6,3,2,1,4,5,8,7,8,9],3)

In [ ]:
## Exercise
## Generate a histogram of the maximum energy of the MAC000113 meter.

#london dataset filtering only one LCLid from MAC000113
meter_113 = london[london['LCLid']=='MAC000113']

# maximum energy of this meter
e_max_113= meter_113[['energy_max']]

In [ ]:
plt.hist(e_max_113.values)

In [ ]:
#plot
e_max_113.hist()

In [ ]:
e_max_113.plot.density()

In [ ]:
import plotly.express as px

In [ ]:
!pip install plotly

## Interactive Plots: Plotly

There are interactive plots, where you can zoom in and out, select data, etc. 

In [ ]:
import plotly.graph_objects as go

# Create traces
fig = go.Figure()
fig.add_trace(go.Scatter(x=meter_113['day'], y=meter_113['energy_max'],
                    mode='lines',
                    name='Energy max - lines'))
fig.add_trace(go.Scatter(x=meter_113['day'], y=meter_113['energy_min'],
                    mode='lines+markers',
                    name='Energy min - lines+markers'))
fig.add_trace(go.Scatter(x=meter_113['day'], y=meter_113['energy_mean'],
                     mode='markers', name='energy_mean - markers'))


# Edit the layout
fig.update_layout(title='Summary ID MAC000113 consumptions',
                   xaxis_title='Month - Year',
                   yaxis_title='Energy (kWh)')


fig.show()